### PR TITLE
Configurable TEID-Ranges

### PR DESCRIPTION
**Pull request by vkatsuba**
Implementation of TDD which allows to define the available prefix of Tunnel Endpoint IDs by configuration.
------------------
Implementation of `TEI` calculation was get from [experimental branch](https://github.com/travelping/ergw/blob/exp/stateless/src/ergw_tei_mngr.erl#L177-L188) what was implemented by @RoadRunnr

### CODE REVIEW

File: src/ergw_node_selection.erl [hunk 1]
```diff
@@ -248,7 +248,7 @@ lookup_host(Name, dns, NameServers) ->
 			      setelement(2, R, [inet_dns:rr(RR, data) | element(2, R)]);
 			  {in, aaaa} ->
 			      setelement(3, R, [inet_dns:rr(RR, data) | element(3, R)]);
-			  Other ->
+			  _Other ->
```
Comment (ebengt) [inline src/ergw_node_selection.erl:251]: It might be worth adding _ prefix to unused variables in a separate commit.

File: .gitignore [hunk 2]
```diff
@@ -11,3 +11,6 @@ rel/example_project
 tetrapak/.local.cache
 *~
 _build
+rebar3
```
Comment (ebengt) [inline .gitignore:14]: It might be worth adding these in a separate commit.

File: README.md [hunk 3]
```diff
@@ -105,36 +105,27 @@ DOCKER IMAGES
...
-*The minimum supported Erlang version is 22.3.4.*
-
-Erlang 23.0.3 is the recommended version.
-
-Using rebar:
-
-    # rebar3 compile
-
-RUNNING
+BUILDING & RUNNING
```
Comment (ebengt) [inline README.md:108]: Good explanation that could perhaps go into a separate commit.

File: README.md [hunk 4]
```diff
@@ -105,36 +105,27 @@ DOCKER IMAGES
...
 -------
-
+Required:
+* The minimum supported Erlang version is **22.3.4**(*Erlang 23.0.3 is the recommended version.*).
```
Comment (jellonek) [inline README.md:111]: Please add space before `(`.

File: src/gtp_context_reg.erl [hunk 5]
```diff
@@ -27,12 +27,15 @@
 %% Include files
 %% --------------------------------------------------------------------
 -include("include/ergw.hrl").
+-include_lib("kernel/include/logger.hrl").
 
 -define(SERVER, ?MODULE).
--record(state, {pids, await_unreg}).
+-record(state, {pids, await_unreg, teid_cfg}).
```
Comment (ebengt) [inline src/gtp_context_reg.erl:33]: The possible interpretations of cfg (as an acronym) are about 100(?). Writing config makes it easier to understand without being too verbose.

File: README.md [hunk 6]
```diff
@@ -105,36 +105,27 @@ DOCKER IMAGES
...
+```sh
+$ git clone https://github.com/travelping/ergw.git
+$ cd ergw
+$ wget https://s3.amazonaws.com/rebar3/rebar3
+$ chmod u+x ./rebar3
+$ touch ergw.config
 ```
 
-This requires a suitable ergw.config, e.g.:
+Then edit the file that just was created **ergw.config** like described below. This requires a suitable **ergw.config**, e.g.:
```
Comment (jellonek) [inline README.md:128]:
```suggestion
then fill just created **ergw.config** file with content like described below providing a suitable configuration, e.g.:
```

File: src/gtp_context_reg.erl [hunk 7]
```diff
@@ -215,14 +219,38 @@ maybe_init_rnd([]) ->
 maybe_init_rnd([{_, RndState}]) ->
     RndState.
 
-alloc_tei(_RndStateKey, _RndState, _KeyFun, 0) ->
+alloc_tei(_RndStateKey, _RndState, _KeyFun, 0, _) ->
     {error, no_tei};
-alloc_tei(RndStateKey, RndState0, KeyFun, Cnt) ->
-    {TEI, RndState} = rand:uniform_s(16#fffffffe, RndState0),
+alloc_tei(RndStateKey, RndState0, KeyFun, Cnt, #{prefix := Prefix0, length := PrefixLen} = Data) ->
```
Comment (ebengt) [inline src/gtp_context_reg.erl:224]: Data was called TeidCfg in the call. It is easier to understand if it keeps this name.

File: src/gtp_context_reg.erl [hunk 8]
```diff
@@ -215,14 +219,38 @@ maybe_init_rnd([]) ->
...
 	_ ->
-	    alloc_tei(RndStateKey, RndState, KeyFun, Cnt - 1)
+	    alloc_tei(RndStateKey, RndState, KeyFun, Cnt - 1, Data)
     end.
+
+get_teid_cfg() ->
+	get_teid_pref_len(application:get_env(ergw, teid), #{prefix => 0, length => 1}).
+
+get_teid_pref_len({ok, [_|_] = TeidCfg}, DefTEID) ->
+	case proplists:get_value(length, TeidCfg, 0) of
```
Comment (ebengt) [inline src/gtp_context_reg.erl:241]: It is easier to read code that does not nest case. Perhaps split this into two help functions?

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
This PR could be seen as precursor to https://github.com/travelping/ergw/commit/f977f61a4a46cf287fb9d39c3d9771ef673247ac.
I would therefore recommend to use a stripped down version of that implementation (remove the Nudsf storage).

File: rebar.config [hunk 9]
```diff
@@ -97,7 +97,9 @@
   % test case, but leave it up to the tests to actually create the directories.
   {create_priv_dir, manual_per_tc},
   % Specify that we want the results captured in a Surefire report.
-  {ct_hooks, [{cth_surefire, [{path, "../report.xml"}]}]}
+  {ct_hooks, [{cth_surefire, [{path, "../report.xml"}]}]},
+  % Config for tests
+  {sys_config, ["test/test.config"]}
```
[the remark this answers is not in the document]
Comment (vkatsuba) [reply rebar.config:102]: Because I see that in CT we have a lot of places where we use the same values like `{'MS-Primary-DNS-Server', {8,8,8,8}}` and others, this is a simple copy and paste, the firs point why I added is - do set configuration for TEID only once and second point is also if we will be have some immutable configuration we can set configuration only once without putting the same one config in a lot of places in CTs. But if you think that this is bad approach I can remove it.

File: src/ergw_tei_mngr.erl [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+get_teid_pref_len({ok, {Prefix, Len}}, _) ->
+	{Prefix, Len};
+get_teid_pref_len({ok, {Len}}, DefaultConfig) when not is_integer(Len); Len < 0; Len > ?MAX_TRIES ->
+	?LOG(warning, "TEID length has wrong configuration: ~p", [Len]),
+	DefaultConfig;
+get_teid_pref_len({ok, {Len}}, _) ->
+	{(fun(X) when X =< 8 -> X; (_) -> 1 end)(?POW(2, Len)), Len};
+get_teid_pref_len(WrongConfig, DefaultConfig) ->
+	?LOG(warning, "TEID has wrong configuration: ~p", [WrongConfig]),
+	DefaultConfig.
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:170]: please look how the other modules implement validate_config methods or implement this in ergw_config.

File: README.md (hunk 4 above)
Comment (vkatsuba) [reply README.md]: Moved README changes into https://github.com/travelping/ergw/pull/214

File: README.md (hunk 6 above)
Comment (vkatsuba) [reply README.md]: Moved README changes into https://github.com/travelping/ergw/pull/214

File: README.md (hunk 3 above)
Comment (vkatsuba) [reply README.md]: Moved README changes into https://github.com/travelping/ergw/pull/214

File: src/ergw_tei_mngr.erl [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	when not is_integer(Len); Len < 0; Len > ?MAX_TRIES; not is_integer(Prefix); Prefix < 0; Prefix > 8 ->
+		?LOG(warning, "TEID length and/or prefix has wrong configuration: ~p", [WrongConfig]),
+		DefaultConfig;
+get_teid_pref_len({ok, {Prefix, Len}}, _) ->
+	{Prefix, Len};
+get_teid_pref_len({ok, {Len}}, DefaultConfig) when not is_integer(Len); Len < 0; Len > ?MAX_TRIES ->
+	?LOG(warning, "TEID length has wrong configuration: ~p", [Len]),
+	DefaultConfig;
+get_teid_pref_len({ok, {Len}}, _) ->
+	{(fun(X) when X =< 8 -> X; (_) -> 1 end)(?POW(2, Len)), Len};
```
Comment (vkatsuba) [inline src/ergw_tei_mngr.erl:167]:
```erlang
(fun(X) when X =< 3 -> ?POW(2, X); (_) -> 1 end)(Len).
```

File: src/ergw_config.erl [hunk 12]
```diff
@@ -119,6 +125,21 @@ validate_node_name(Name)
...
+	{Prefix, Len};
+validate_teid({ok, {Len}}, DefaultConfig) when not is_integer(Len); Len < 0; Len > ?MAX_TRIES ->
+	?LOG(warning, "TEID length has wrong configuration: ~p", [Len]),
+	DefaultConfig;
+validate_teid({ok, {Len}}, _) ->
+	{(fun(X) when X =< 3 -> ?POW(2, X); (_) -> 1 end)(Len), Len};
+validate_teid(WrongConfig, DefaultConfig) ->
+	?LOG(warning, "TEID has wrong configuration: ~p", [WrongConfig]),
+	DefaultConfig.
+
```
Comment (RoadRunnr) [inline src/ergw_config.erl:142]: validations happens before application startup and not later.
Please make this part of the already existing validation runs.

File: src/ergw_config.erl [hunk 13]
```diff
@@ -268,6 +268,8 @@ validate_option(Opt, Value)
        Opt == proxy_map;
        Opt == path_management ->
     throw({error, {options, {Opt, Value}}});
+validate_option(teid, Value) ->
+    ergw_tei_mngr:validate_options({teid, Value});
```
Comment (RoadRunnr) [inline src/ergw_config.erl:272]: that's not going to work

File: src/ergw_tei_mngr.erl [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+get_teid_config() ->
+    case application:get_env(ergw, teid) of
+        {ok, {_, _} = Config} ->
+            Config;
+        {ok, {Len}} ->
+            {?POW(2, Len), Len};
+        _ ->
+            {1, 0}
+    end.
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:174]: validation phase should take care of that. The init method should be able to assume that the config key always contains sensible data without this case here.

File: rebar.config (hunk 9 above)
[the remark this answers is not in the document]
Comment (RoadRunnr) [reply rebar.config:102]: Yes. Please, remove that config file.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
test case for the config validation in config_SUITE is missing.

File: src/ergw_tei_mngr.erl (hunk 14 above)
Comment (vkatsuba) [reply src/ergw_tei_mngr.erl:174]: We have a 3 cases:
1 - when set Length and Prefix
2 - when set Length only
3 - when configuration is now a set at all
Here we don't do validation, we try to create default value for Prefix set default Length and Prefix

File: src/ergw_config.erl (hunk 13 above)
Comment (vkatsuba) [reply src/ergw_config.erl:272]: Can you share - why? It's was tested with some fail config data and looks like validation is working as expected.

Comment (RoadRunnr) [reply src/ergw_config.erl:272]: validate_options expects a list or map, you are passing a tuple

File: src/ergw_tei_mngr.erl (hunk 14 above)
Comment (RoadRunnr) [reply src/ergw_tei_mngr.erl:174]: your implementation logic always expect a tuple `{Length, Prefix}`. And that is all it should care about. The logic of having a default or just a single number, should go into the validation step.

File: src/ergw_config.erl (hunk 13 above)
Comment (vkatsuba) [reply src/ergw_config.erl:272]: `ergw_tei_mngr:validate_options/1` expect tuple, and then try put a list in `ergw_config:validate_options/4` and call looks like `ergw_config:validate_options(fun validate_option/1, [{teid, {Len, Prefix}}], [], list).`. So, into `ergw_config:validate_option/4` is set the list.

File: src/ergw_tei_mngr.erl [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+    TID = ets:new(?SERVER, [ordered_set, {keypos, 1}]),
+    NodeId = node(),
+    BlockId = iolist_to_binary(["teid-prefix-", integer_to_list(Prefix)]),
+
+    Data = #data{node_id = NodeId,
+		  block_id = BlockId,
+		  prefix = Prefix,
+		  prefix_len = PrefixLen,
+		  tid = TID,
+		  save_cnt = ?SAVE_COUNTER},
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:99]: node_id, block_id and save_cnt are not really use in this version. Even when we later use them, adding them in this version is not needed.

File: src/ergw_tei_mngr.erl [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+serialize_data(State, #data{tid = TID}) ->
+    SeedL = maps:fold(fun(K,V,L) -> [{K, rand:export_seed_s(V)}|L] end, [], State),
+    Term =
+	#{seed  => SeedL,
+	  teids => ets:tab2list(TID)},
+    term_to_binary(Term, [{minor_version, 2}, compressed]).
+
+save_data(State, #data{node_id = _NodeId, block_id = _BlockId} = Data) ->
+    _Serialized = serialize_data(State, Data).
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:149]: state_timeout, serialize_data and save_data are not really used in this version.

File: src/ergw_tei_mngr.erl [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+alloc_tei_reply(From, Reply, Name, RndState, State, Data) ->
+    Actions = [{reply, From, Reply}, state_timeout()],
+    {next_state, State#{Name => RndState}, Data, Actions}.
+
+alloc_tei(From, Name, _KeyFun, 0, RndState, State, Data) ->
+    alloc_tei_reply(From, {error, no_tei}, Name, RndState, State, Data);
+alloc_tei(From, Name, KeyFun, Cnt, RndState0, State,
+	  #data{prefix = Prefix0, prefix_len = PrefixLen, tid = TID} = Data) ->
+    Prefix = Prefix0 bsl (?MAX_TRIES - PrefixLen),
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:159]: ?MAX_TRIES is wrong, that it is 32 is just a coincidence.

File: src/ergw_tei_mngr.erl [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+-export([callback_mode/0, init/1, handle_event/4, terminate/3, code_change/4]).
+
+-export([validate_option/1]).
+
+-include("include/ergw.hrl").
+
+-define(SERVER, ?MODULE).
+-define(SAVE_COUNTER, 10).
+-define(SAVE_TIMEOUT, 1000).
+-define(MAX_TRIES, 32).
```
Comment (RoadRunnr) [inline src/ergw_tei_mngr.erl:31]: 32 is actually quite high for retries, 10 should be by far enough.

File: test/ggsn_SUITE.erl [hunk 19]
```diff
@@ -154,6 +154,8 @@
 		     {ip_pools, ['pool-A']}]}
 		  ]},
 
+		 {teid, {3, 6}},
+
 		 {charging,
```
Comment (RoadRunnr) [inline test/ggsn_SUITE.erl:159]: does the test work without that? the default for teid should have worked (but I didn't test that)

**Review by RoadRunnr — CHANGES_REQUESTED** [review]
there are other things that originate from the Nudsf version of the teid manager that are not needed in this version. They should not be in here.

File: test/ggsn_SUITE.erl (hunk 19 above)
Comment (vkatsuba) [reply test/ggsn_SUITE.erl:159]: Without this setting `{3, 6}` the tests is failed

Comment (RoadRunnr) [reply test/ggsn_SUITE.erl:159]: it should work without, the default is {0,0}.
Can you find out why it failes?

Comment (vkatsuba) [reply test/ggsn_SUITE.erl:159]: Looks like the value `{0, 0}` is not set by default.

Comment (vkatsuba) [reply test/ggsn_SUITE.erl:159]: The default value cannot be set by reason that in `ergw_app:start/2` we call `Pid <- ergw_sup:start_link(),`(where we start `ergw_tei_mngr` - when we call  `ergw_tei_mngr` and try get in the function `init` the configurations, the config is not exist and default value is not set) and then we call `ergw_config:load_config(setup:get_all_env(ergw))` where we are set the default value in configuration. Plus we need also update `ergw:load_config/1` for set default value for new options. Not sure, but looks like the `init` function is not a best plays for getting config of `Prefix, PrefixLen` in current case.

Comment (vkatsuba) [reply test/ggsn_SUITE.erl]: Fixed.

**Review by RoadRunnr — APPROVED** [review]
please squash the changes into a single commit before merging.
